### DISCUSSION TITLE
some errors in user subroutine

### DISCUSSION THREAD

**Discussion by LYS0922** [OpenRadioss/OpenRadioss#3000] [Q&A]
some errors in user subroutine
1、the function GET_U_MAT(4,IMAT) is invalid even in the case you released .(triangle shell element , sini29.f , and suser29.f.)
2、the initial density RHO0(I)  become 0, in the loop of first step.

Is anyone interested in secondary development and user subroutines here
We can communicate

**Comment by LYS0922** [comment]
It seems that few people use user's subroutine

**Reply by MarianBulla** [reply]
Hi @LYS0922,
thank you very much for using OpenRadioss and the user subroutines for developing your own code.

Is it possible to get your source code (or a simplified version of it) to check where the issue is, please ?
Also with a small model. One element test should be sufficient enough.

There are several developments ongoing with the usage of the user subroutines and the sdk.
So, people are using it, which makes your issue very important and relevant.

Thanks and best regards,
Marian

**Reply by LYS0922** [reply]
Hi@MarianBulla
Very thank your attention.
I will show the problem for you by giving the source and pointed out the failure codes.
Firstly, we can get the user's subroutine Example Demo File from https://help.altair.com/hwsolvers/rad/topics/solvers/rad/user_property_elements_triangular_shell_elements_solid_property_example.htm

Then, compile it, run the one element model.  
In the starter print part ,there are errors.
"ERROR ID :    376
** ERROR IN USER SET (PARAMETER INDEX)
DESCRIPTION :
   USER  PARAMETER INDEX(0),
   EXCEED (0)"
This error is because the failure of function "IMAT = GET_U_PNU(1,IPROP,KMAT)"  in line76 in sini29.f.
IMAT should be 1，however it is 2.
You could insert this codes afer line 76 to see and change the value of IMAT.
"         print *, 'IMAT =',IMAT 
	  IMAT = 1
	  print *, 'IMAT =',IMAT "
After inserting this codes, it could initialize successfully.
In the engine print part ,there are errors.
  **ERROR USER MATERIAL PARAMETER INDEX(           1 ) EXCEED NUPARAM(
           0 )
This is also because the the failure of function "IMAT = GET_U_PNU(1,IPROP,KMAT)"  in line 315 and 368 in suser29.f
By inserting the codes to change the value of IMAT to 1, this model could run successfully.
So, I guess this problem is caused by an error in the a function.   “GET_U_PNU(1,IPROP,KMAT)”
Look forward your reply.
LYS

**Reply by LYS0922** [reply]
Hi, @MarianBulla 
Apart from the previous issues， I still have 2 problems  which is pointed out in the attached file  sigeps29.f , and 1 question.
[for_show.zip](https://github.com/user-attachments/files/17101562/for_show.zip)
This is  a cube model with 125 element.
The problem 1 is occured in line 64 to 84 in sigeps29.f that the value of UPARAM(*)  is loss.
The problem 2 is occured in line 101 to 106 in sigeps29.f that the value of RHO0(I) and SOUNDSP(I)  is loss.
I speculate that these two problems prevent the solver from proceeding to the next time step.
Question:
What version of FORTRAN does the user's subroutine use？

Sincerely looking forward to your answer.
LYS
